client/js/contextMenuFactory.js

```diff
@@ -0,0 +1,193 @@
+	}
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("user"),
+		className: "user",
+		display: (target) => target.text(),
+		data: (target) => target.data("name"),
+		divider: true,
+		callback: whois,
+	});
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("user"),
+		className: "action-whois",
+		display: "User information",
+		data: (target) => target.data("name"),
+		callback: whois,
+	});
+}
+
+function addQueryItem() {
+	function query(itemData) {
+		const chan = utils.findCurrentNetworkChan(itemData);
+
+		if (chan.length) {
+			chan.click();
+		}
+
+		socket.emit("input", {
+			target: $("#chat").data("id"),
+			text: "/query " + itemData,
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("user"),
+		className: "action-query",
+		display: "Direct messages",
+		data: (target) => target.data("name"),
+		callback: query,
+	});
+}
+
+function addKickItem() {
+	function kick(itemData) {
+		socket.emit("input", {
+			target: $("#chat").data("id"),
+			text: "/kick " + itemData,
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) => utils.hasRoleInChannel(target.closest(".chan"), ["op"]) && target.closest(".chan").data("type") === "channel",
+		className: "action-kick",
+		display: "Kick",
+		data: (target) => target.data("name"),
+		callback: kick,
+	});
+}
+
+function addFocusItem() {
+	function focusChan(itemData) {
+		$(`.networks .chan[data-target="${itemData}"]`).click();
+	}
+
+	const getClass = (target) => {
+		if (target.hasClass("lobby")) {
+			return "network";
+		} else if (target.hasClass("query")) {
+			return "query";
+		}
+
+		return "chan";
+	};
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("chan"),
+		className: getClass,
+		display: (target) => target.attr("aria-label"),
+		data: (target) => target.data("target"),
+		divider: true,
+		callback: focusChan,
+	});
+}
+
+function addChannelListItem() {
+	function list(itemData) {
+		socket.emit("input", {
+			target: itemData,
+			text: "/list",
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("lobby"),
+		className: "list",
+		display: "List all channels",
+		data: (target) => target.data("id"),
+		callback: list,
+	});
+}
+
+function addBanListItem() {
+	function banlist(itemData) {
+		socket.emit("input", {
+			target: itemData,
+			text: "/banlist",
+		});
+	}
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("channel"),
+		className: "list",
+		display: "List banned users",
+		data: (target) => target.data("id"),
+		callback: banlist,
+	});
+}
+
+function addJoinItem() {
+	function openJoinForm(itemData) {
+		const network = $(`#join-channel-${itemData}`).closest(".network");
+		JoinChannel.openForm(network);
+	}
+
+	addContextMenuItem({
+		check: (target) => target.hasClass("lobby"),
+		className: "join",
+		display: "Join a channel…",
+		data: (target) => target.data("id"),
+		callback: openJoinForm,
+	});
+}
+
+function addDefaultItems() {
+	addWhoisItem();
```

Comment:
So, this now just calls each of the functions. I'm not 100% on this just now. Ideally, each of these functions are called somewhere else (next to code for this feature) but ... that doesn't really exist just now for most of it. So we'll keep this all here for now, but ... eventually it'll be nicer, I think.